---
title: "Shalek PhenoPath analysis"
author: "Kieran Campbell"
date: "September 2017"
output: html_document
---

```{r setup, include=FALSE, cache=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = FALSE, fig.width = 6, fig.height = 4)

library(scater)
library(tidyverse)
library(magrittr)
library(matrixStats)
library(ggrepel)
library(cowplot)
library(viridis)
library(splines)
library(DT)
library(limma)
library(edgeR)
library(RColorBrewer)
library(goseq)
library(phenopath)
library(ggbeeswarm)



# http://stackoverflow.com/questions/18509527/first-letter-to-upper-case
firstup <- function(x) {
   substr(x, 1, 1) <- toupper(substr(x, 1, 1))
x
}

set.seed(123L)

scale_vec <- function(x) (x - mean(x)) / sd(x)

```

# Fit the CLVM model

```{r}
sce <- readRDS("../../data/paper-scesets/sce_shalek_qc.rds")
sce <- updateSCESet(sce)
gene_vars <- rowVars(exprs(sce))
var_cutoff <- sort(gene_vars, decreasing = TRUE)[7500]
sce_hvg <- sce[gene_vars >= var_cutoff, ]

is_na_mgi_symbol <- is.na(rowData(sce_hvg)$mgi_symbol)
rowData(sce_hvg)$mgi_symbol[is_na_mgi_symbol] <- rownames(sce_hvg)[is_na_mgi_symbol]

is_Rasgefb1 <- match("ENSMUSG00000029333.14", rownames(sce_hvg))
rowData(sce_hvg)$mgi_symbol[is_Rasgefb1] <- "Rasgefb1"
```

And perform inference:

```{r}
pc1 <- prcomp(t(exprs(sce_hvg)), scale = TRUE)$x[,1]
pc1 <- scale_vec(pc1)
time_numeric <- as.numeric(gsub("h", "", sce$time))
pc1 <- pc1 * sign(cor(pc1, time_numeric))
x <- 2 * (sce_hvg$stimulant == "LPS") - 1

pcavi <- phenopath(sce_hvg, x, z_init = pc1, sce_assay = "norm_exprs")
```

Save for decision analysis:

```{r}
dec_df <- data_frame(gene = rownames(sce_hvg),
                     m = pcavi$m_beta[1,],
                     s = pcavi$s_beta[1,])
write_csv(dec_df, "../../data/shalek/dec_df_new.csv")
```


Plot the elbo:

```{r plot-elbo}
plot_elbo(pcavi) +
  xlab(expression("CAVI iterations")) +
  theme_cowplot(font_size = 11)

ggsave("../../figs/elbo_shalek.png", width = 5, height = 3)
```


# Exploring PhenoPath results

## Comparison to PC1

```{r pseudotime-vs-pc1}
xs <- sce_hvg$stimulant
 
qplot(scale_vec(pc1), pcavi$m_z, color = factor(xs)) +
  scale_color_brewer(palette = "Set1", name = "Stimulant") +
  #stat_function(fun = function(x) x, color = 'black') + 
  xlab("PC1") + ylab(expression(z))
```


## Identifying significant genes


```{r beta-df}

rowData(sce_hvg)$ensembl_gene_id <- sapply(strsplit(rownames(sce_hvg), ".", fixed = TRUE), `[`, 1)

df_beta <- data_frame(
  lambda = pcavi$m_lambda,
  beta = pcavi$m_beta[1,],
  chi = pcavi$chi_exp[1,],
  alpha = pcavi$m_alpha[1,],
  pos_sd = sqrt(pcavi$s_beta[1,]),
  gene = rowData(sce_hvg)$mgi_symbol, #stringr::str_to_title(rownames(sce_hvg)),
  ensembl_gene_id = rowData(sce_hvg)$ensembl_gene_id,
  is_sig = significant_interactions(pcavi, n = 3)#[,1]
)


write_csv(df_beta, "../../data/shalek/df_gene_shalek.csv")


df_beta$is_sig_graph <- 
  plyr::mapvalues(df_beta$is_sig, from = c(FALSE, TRUE),
                  to = c("Non-significant", "Significant"))
```

This gives:

```{r how-many-sig}
sum(df_beta$is_sig)
```

_significant_ interactions.

We can plot these:

```{r plot-sigs}
chi_cutoff <- 0.15
df_beta <- dplyr::mutate(df_beta, 
                  gene_symbol = gene)


ggplot(df_beta, aes(x = beta, y = 1 / chi, color = is_sig_graph)) + 
  geom_point(alpha = 0.8) +
  theme(legend.title = element_blank()) +
  # geom_point(data = dplyr::filter(df_beta, is_mmr), alpha = 0.6) +
  scale_color_brewer(palette = "Set1", name = "'Significant'") +
     geom_label_repel(data = dplyr::filter(df_beta, 1 / chi > chi_cutoff), aes(label = gene_symbol)) +
  xlab(expression(beta)) +
  ylab(expression(paste("[", chi, "]"^-1)))
  
```

Which genes are sig?

```{r dt-sig}
dplyr::filter(df_beta, is_sig) %>% dplyr::arrange(desc(abs(beta))) %>% 
  dplyr::select(gene, beta, alpha) %>% 
  datatable()
```

We can plot expression along phenotime for the 35 _significant_ genes:

```{r exprs-along-pst, fig.width = 12, fig.height = 9}
tmap <- pcavi$m_z

top_genes <- dplyr::filter(df_beta, is_sig) %>% 
  dplyr::arrange(desc(abs(beta))) %>% 
  extract2("gene") %>% head(n=20)
```

Check z loosely corresponds to time:

```{r dfz}
zdf <- data_frame(z = pcavi$m_z, time = sce$time, stimulant = xs)

ggplot(zdf, aes(x = time, y = z, fill = time)) + 
  geom_violin(alpha = 0.8) +
  theme(legend.position = "none") + 
  scale_fill_brewer(palette = "Set3") +
  # scale_fill_brewer(palette = "Set1") +
  xlab("Capture time") +
  ylab("Pseudotime)") +
  theme(axis.text = element_text(size = 8),
        axis.title = element_text(size = 9),
        legend.title = element_text(size = 11),
        legend.text = element_text(size = 10)) 

time_plot <- last_plot()
```




```{r mlh1}

make_gene_plot <- function(genes) {
  inds <- match(genes, rowData(sce_hvg)$mgi_symbol)
  
  df <- t(exprs(sce_hvg))[, inds, drop=FALSE] 
  colnames(df) <- genes #rownames(sce_hvg)[inds]
  
  df <- as_data_frame(df) %>% 
    dplyr::mutate(phenopath = pcavi$m_z, x = as.character(xs)) %>% 
    gather(gene, expression, -phenopath, -x)
  
  df$gene <- factor(df$gene, levels = genes)
  
  df_nox <- dplyr::select(df, -x)
  
  ggplot(df, aes(x = phenopath, y = expression, fill = x, color = x)) + 
    geom_point(data = df_nox, color = "grey80", fill = "grey80") +
    geom_point(shape = 21, color = 'grey30', alpha = 0.3) +
    facet_grid(gene ~ x, scales = "free") + 
    scale_fill_brewer(palette = "Set1", name = "MSI") +
    scale_color_brewer(palette = "Set1", name = "MSI") +
    theme(legend.position = "none", 
          strip.text = element_text(face = "bold"),
          strip.text.x = element_text(size = 10, margin = margin(.1, 0, .1, 0, "cm")),
          strip.text.y = element_text(size = 8, margin = margin(0, .1, 0, .1, "cm")),
          strip.background = element_rect(colour = "black", fill = "grey95", 
                                          size = 0.3, linetype = 1)) +
  theme(axis.text = element_text(size = 9),
        axis.text.y = element_text(size = 8),
        axis.title = element_text(size = 10)) +
    stat_smooth(se = FALSE, method = "lm", size = 1.5, formula = y ~ ns(x,3),
                color = 'grey30') +
    ylab(expression("Normalised expression")) + 
    xlab("Pseudotime")
}

# genes <- c("Ifi204", "Ms4a6c", "Procr")
genes <- top_genes[1:4]

make_gene_plot(genes)

gene_plot <- plot_over_pseudotime <- last_plot()
```

Now do the same for original time series:

```{r}
inds <- match(genes, rowData(sce_hvg)$mgi_symbol)

df <- t(exprs(sce_hvg))[, inds, drop=FALSE] 
colnames(df) <- genes
  
df <- as_data_frame(df) %>% 
    dplyr::mutate(capture_time = sce_hvg$time, x = as.character(xs)) %>% 
    gather(gene, expression, -capture_time, -x)
  
df$gene <- factor(df$gene, levels = genes)

df_g <- group_by(df, capture_time, x, gene) %>% 
  summarise(median_expression = median(expression))

ggplot(df, aes(x = capture_time, y = expression, fill = x, color = x)) + 
    # geom_beeswarm(shape = 21, color = 'grey30', alpha = 0.3) +
    geom_jitter(shape = 21, color = 'grey30', alpha = 0.3, width= 0.1) +
    facet_grid(gene ~ x, scales = "free") + 
    scale_fill_brewer(palette = "Set1") +
    scale_color_brewer(palette = "Set1") +
    theme(legend.position = "none", 
          strip.text = element_text(face = "bold"),
          strip.text.x = element_text(size = 10, margin = margin(.1, 0, .1, 0, "cm")),
          strip.text.y = element_text(size = 8, margin = margin(0, .1, 0, .1, "cm")),
          strip.background = element_rect(colour = "black", fill = "grey95", 
                                          size = 0.3, linetype = 1)) +
  theme(axis.text = element_text(size = 9),
        axis.text.y = element_text(size = 8),
        axis.title = element_text(size = 10)) +
    ylab(expression("Normalised expression")) + 
    xlab("Capture time") +
      geom_point(data = df_g, aes(y = median_expression), 
                 shape = 21, color = 'black', size = 2) +
  geom_line(data = df_g, aes(y = median_expression, group = 1))

plot_over_capture_time <- last_plot()

# ggsave("../../figs/supplementary_time_shalek.png", width = 6, height = 4)
```

We have everything needed for plot 1:

```{r}
other_pseudotime_comparison_plot <- readRDS("../../figs/shalek_cor.rds")
lsize = 12
upper_grid <- plot_grid(time_plot, other_pseudotime_comparison_plot,
                        rel_widths = c(1,2), labels = c("A", "B"),
                        label_size = lsize)
bottom_grid <- plot_grid(plot_over_capture_time, plot_over_pseudotime,
                         labels = c("C", "D"),
                         label_size = lsize)

plot_grid(upper_grid, bottom_grid, nrow = 2,
          rel_heights = c(2,3))

ggsave("../../figs/new/shalek_2.png", width = 9, height = 6)
```



# Differential expression comparison

We'll use limma-voom for standard differential expression:


```{r limma}
dge <- DGEList(counts(sce_hvg))
dge <- calcNormFactors(dge)

design <- model.matrix(~ x, colData(sce_hvg))
v <- voom(dge, design, plot = TRUE)

fit <- lmFit(v, design)
fit <- eBayes(fit)
results <- decideTests(fit)


df_limma <- data_frame(coef = fit$coefficients[,2], 
                       pval = fit$p.value[,2],
                       mu = pcavi$m_mu,
                       ensembl_gene_id = rowData(sce_hvg)$ensembl_gene_id) %>% 
  left_join(df_beta, by = "ensembl_gene_id") %>% 
  dplyr::mutate(qval = p.adjust(pval, method = "BH"),
                log10qval = -log10(qval))
```


```{r compare-beta}

cols <- RColorBrewer::brewer.pal(3, "Set2")
cols2 <- c("#c5e2d9", cols[2])

ggplot(df_limma, aes(x = beta, y = log10qval, color = is_sig_graph)) + 
  geom_point() +
  ylab(expression(paste("Limma Voom -", log[10], "(q-value)"))) + 
  xlab(expression(paste("PhenoPath ", beta))) +
  geom_hline(yintercept = -log10(0.05), linetype = 2, alpha = 0.5) +
  theme(axis.text = element_text(size = 8),
        axis.title = element_text(size = 9),
        axis.title.y = element_text(size = 8)) +
  theme(legend.title = element_text(size = 9),
        legend.text = element_text(size = 8)) +
  scale_color_manual(values = cols2, name = "Interaction") +
  theme(legend.margin = margin(), legend.position = "none")

de_plot <- ggExtra::ggMarginal(last_plot(), margins = "y", type = "histogram", size = 10)
```

Again no obvious relationship.

# Crossover bit for chris

```{r df-beta}
df_beta <- mutate(df_beta,
                  crossover = - alpha / beta)

filter(df_beta, is_sig) %>% 
  ggplot(aes(x = crossover)) + geom_histogram()
```


```{r save-supp-data}
write_csv(df_beta, "../../data/shalek/shalek_interactions.csv")
```

```{r more-vis}
textinfo <- frame_data(
  ~x, ~y, ~label,
  1.2, 0.2, "Gene upregulated along pseudotime\nLPS increases upregulation",
  1.2, -0.2, "Gene upregulated along pseudotime\nPAM increases upregulation",
  -1.3, 0.2, "Gene downregulated along pseudotime\nPAM increases downregulation",
  -1.3, -0.2, "Gene downregulated along pseudotime\nLPS increases downregulation"
)

cols <- RColorBrewer::brewer.pal(3, "Set2")
cols2 <- c("#c5e2d9", cols[2])

outline_cols = c("#c5e2d9", 'black')

ggplot(df_beta, aes(x = lambda, y = beta)) +
  geom_point(shape = 21, aes(fill = is_sig_graph, color = is_sig_graph)) +
  geom_vline(xintercept = 0, linetype = 2, alpha = 0.5) +
  geom_hline(yintercept = 0, linetype = 2, alpha = 0.5) +
  scale_fill_manual(values = cols2, name = "Interaction") +
  scale_color_manual(values = outline_cols, name = "Interaction") +
  geom_text_repel(data = dplyr::filter(df_beta, (beta > 0.8 & is_sig) | beta < -1.3),
                aes(label = gene), color = 'black',
                size = 3) +
  ylab("Covariate-pseudotime interaction") +
  xlab("Gene regulation over pseudotime") +
  theme(legend.position = 'bottom',  #, legend.margin = margin(),
        axis.text = element_text(size = 10),
        axis.title = element_text(size = 11),
        legend.title = element_text(size = 11),
        legend.text = element_text(size = 10)) +
  geom_text(data = textinfo, aes(x = x, y = y, label = label), 
            color = 'black', size = 3, fontface = "bold")
lplot <- ggExtra::ggMarginal(last_plot(), margins = "y", type = "histogram", size = 10)
print(lplot)
```

# What is pseudotime (2)


```{r get-correlations}
genome <- "mm9"
id <- "ensGene"

all_genes <- rowData(sce_hvg)$ensembl_gene_id

upreg_genes <- filter(df_beta, is_sig, beta > 0) %>% extract2("ensembl_gene_id") 
downreg_genes <- filter(df_beta, is_sig, beta < 0) %>% extract2("ensembl_gene_id")

upg <- 1 * (all_genes %in% upreg_genes)
downg <- 1 * (all_genes %in% downreg_genes)

names(upg) <- names(downg) <- all_genes

pwfup <- nullp(upg, genome, id)
goup <- goseq(pwfup, genome, id, test.cats = "GO:BP")

pwfdown <- nullp(downg, genome, id)
godown <- goseq(pwfdown, genome, id, test.cats = "GO:BP")
```

Graph results:

```{r graph-res}
parse_go <- function(go, type, n_tested) {
  go <- go %>% 
    mutate(qval = p.adjust(over_represented_pvalue)) %>% 
    mutate(log10qval = -log10(qval),
           prop_in_cat = numInCat / n_tested) %>% 
    head(n = 6) %>% 
    mutate(type = type) %>% 
    tbl_df() %>% 
    arrange(desc(log10qval))
  go
}


gos <- bind_rows(
  parse_go(goup, "LPS up-regulated", length(upreg_genes)),
  parse_go(godown, "PAM up-regulated", length(downreg_genes))
)

# gos <- filter(gos, type == "LPS up-regulated") 

gos$term <- stringr::str_to_title(gos$term)
gos$term <- factor(gos$term, levels = gos$term[order(gos$log10qval)])

gos <- mutate(gos, prop_cat = numDEInCat / numInCat)


gos %>% 
  filter(qval < 0.05) %>% 
ggplot(aes(x = term, y = log10qval)) + #, size = 100 * prop_cat)) +
  geom_segment(aes(y = min(log10qval - 1), yend = log10qval, x = term, xend = term),
               color = 'grey30', linetype = 3) +
    geom_point(aes(color = type, shape = type)) +
    coord_flip() +
  # facet_wrap(~ type, ncol = 1, scales = "free") +
    theme(axis.title.y = element_blank()) +
    ylab(expression(paste("-", log[10], " q-value"))) +
  scale_size(name = "% category\nrepresented") +
    theme(legend.position = "bottom", #c(0.6,0.1),
          legend.direction = "vertical",
        axis.text = element_text(size = 8),
        axis.text.y = element_text(size = 7),
        axis.title = element_text(size = 10),
        legend.title = element_blank(),
        legend.text = element_text(size = 10)) +
  scale_color_brewer(palette = "Set1") +
  theme(strip.text = element_text(size = 8, face = "bold"),
          strip.background = element_rect(colour = "black", fill = "grey95", 
                                          size = 0.5, linetype = 1))
ggsave("../../figs/go.png", width = 5, height = 5)
goplot <- last_plot()
```


First plot for new arrangement:

```{r}
lsize <- 12
right_grid <- plot_grid(de_plot, goplot, label_size = lsize,
                         nrow = 2, labels = c("B", "C"))
main_plot_1 <- plot_grid(lplot, right_grid, labels = c("A", ""), label_size = lsize,
                         ncol = 2, rel_widths = c(2,1))
ggsave(plot = main_plot_1, file = "../../figs/new/shalek_1.png", width = 11, height = 6)
```


```{r large-plot}
lsize <- 11

bottom_top_left <- plot_grid(time_plot,NULL, de_plot, nrow = 1,
                                rel_widths = c(7,1,6),
                                # labels = c("D", "E", ""), 
                             label_size = lsize)

bottom_left <- plot_grid(bottom_top_left, goplot,
                         ncol = 1, # labels = c("", "F"),
                         label_size = lsize)

bottom_grid <- plot_grid(bottom_left, gene_plot, nrow = 1, # labels = c("", "G"),
                         label_size = lsize, rel_widths = c(5.5,5))

shalek_plot <- plot_grid(lplot, bottom_grid, ncol = 1, rel_heights = c(2,1.5), # labels = c("C",""),
          label_size = lsize)

ggsave("../../figs/shalek.png", shalek_plot, width = 8, height = 9)
```

Need a supplementary Mef2c figure

```{r}
mef2c_plot <- make_gene_plot("Mef2c")
ggsave("../../figs/supp_mef2c.png", width = 7, height = 3)
```


# Antiviral pathway score

```{r}
library(readxl)

supp_paper <- read_excel("../../data/supplementary/nature13437-s4.xls")

anti_viral <- mutate(supp_paper, gene = stringr::str_to_title(GENE)) %>% 
  filter(CLUSTER == "Id")

mm <- match(anti_viral$gene, rowData(sce_hvg)$mgi_symbol)
rowData(sce_hvg)$anti_viral <- rep(FALSE, nrow(sce_hvg))
rowData(sce_hvg)$anti_viral[mm] <- TRUE

sce_av <- sce_hvg[rowData(sce_hvg)$anti_viral, ]

scaled_exprs <- scale(t(exprs(sce_av)))

qplot(pcavi$m_z, rowMeans(scaled_exprs), color = xs) +
  labs(x = "Pseudotime", y = "Antiviral score") +
  scale_colour_brewer(palette = "Set1", name = "Stimulant")

ggsave("../../figs/supp_shalek_antiviral_score.png", width = 6, height = 4)
```

# Peaked inflammatory module

```{r}
ints <- interactions(pcavi)
ints$mgi_symbol <- rowData(sce_hvg)$mgi_symbol

peak_genes <- dplyr::filter(supp_paper, CLUSTER == "IIIc") %>% 
  .$GENE %>% 
  stringr::str_to_title()

peak_genes <- dplyr::filter(supp_paper, CLUSTER == "IIIc") %>% 
  .$GENE %>% 
  stringr::str_to_title()

get_intersect_length <- function(n_top) {
  top_ints <- dplyr::arrange(ints, interaction_effect_size) %>% 
    head(n = n_top) %>% 
    .$mgi_symbol
  length(intersect(top_ints, peak_genes)) / n_top
}

get_intersect_length(30)

```

Boxplot for Chris:

```{r}
genes <- c("Tnf", "Rasgefb1", "Tnfaip3", "Malt1")
mm <- match(genes, rowData(sce_hvg)$mgi_symbol)

df_bp <- as_data_frame(t(exprs(sce_hvg)[mm, ]))
names(df_bp) <- genes
df_bp_tidy <- mutate(df_bp, Stimulant = sce_hvg$stimulant) %>% 
  gather(gene, expression, -Stimulant)

ggplot(df_bp_tidy, aes(x = gene, y = expression, fill = Stimulant)) +
  geom_violin() +
  scale_fill_brewer(palette = "Set1") +
  labs(x = "Gene", y = "Normalised Expression")
ggsave("../../figs/shalek_violinplot.png", width = 8, height = 4)
```



# Technical

```{r technical}
sessionInfo()
```
