---
title: "Immune-MSI interactions in colorectal cancer using Variational Bayes PhenoPath"
author: "Kieran Campbell"
date: "February 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = FALSE, fig.width = 6, fig.height = 4)

library(scater)
library(tidyverse)
library(magrittr)
library(matrixStats)
library(ggrepel)
library(cowplot)
library(viridis)
library(splines)
library(DT)
library(edgeR)
library(limma)
library(gplots)
library(RColorBrewer)
library(clvm)

set.seed(123L)

scale_vec <- function(x) (x - mean(x)) / sd(x)

```

# Load data

Load results:

```{r load-all-data}
pcavi <- readRDS("../../data/COAD/clvm_results.rds")
sce <- readRDS("../../data/COAD/sce_coad_clvm.rds")

x <- sce$x
```



Plot the elbo:

```{r plot-elbo}
plot_elbo(pcavi) +
  xlab(expression("CAVI iterations"~x10^{-1})) +
  theme_cowplot(font_size = 11)

ggsave("../../figs/elbo_coad.png", width = 5, height = 3)
```

# Exploring phenotime results

## Comparison to PC1

```{r pseudotime-vs-pc1}
pc1 <- prcomp(t(exprs(sce)), scale = TRUE)$x[,1]

qplot(scale_vec(pc1), scale_vec(pcavi$m_t), color = factor(x)) +
  scale_color_brewer(palette = "Set1", name = "MSI-high") +
  stat_function(fun = function(x) x, color = 'black') + 
  xlab("PC1") + ylab(expression(z))
```

## Identifying significant genes

In the context of CAVI we have modelled each interaction coefficient by an approximating distribution of the form

$$ \beta_{pg} \sim \text{Normal}(m_{\beta_{pg}}, s_{\beta_{pg}}^2) $$
We define a _significant_ interaction if the $2 \sigma$ posterior doesn't cross zero, ie

$$ m_{\beta_{pg}} - n s_{\beta_{pg}} > 0 $$
or
$$ m_{\beta_{pg}} + n s_{\beta_{pg}} < 0 $$

Here we use $n=3$:

```{r beta-df}
df_beta <- data_frame(
  c = pcavi$m_c,
  beta_msi = pcavi$m_beta[1,],
  chi_msi = pcavi$chi_exp[1,],
  alpha_msi = pcavi$m_alpha[1,],
  pos_sd = sqrt(pcavi$s_beta[1,]),
  gene = featureNames(sce),
  is_mmr = fData(sce)$is_mmr,
  is_sig = as.vector(significant_interactions(pcavi, n = 3)),
  hgnc_symbol = fData(sce)$hgnc_symbol,
  ensembl_gene_id = fData(sce)$ensembl_gene_id
)

write_csv(df_beta, "../../data/COAD/df_gene_coad.csv")

df_beta$is_sig_graph <- 
  plyr::mapvalues(df_beta$is_sig, from = c(FALSE, TRUE),
                  to = c("Non-significant", "Significant"))
```

This gives:

```{r how-many-sig}
sum(df_beta$is_sig)
```

_significant_ interactions.

We can plot these:

```{r plot-sigs}
chi_cutoff <- 0.11
df_beta <- mutate(df_beta, 
                  gene_symbol = sapply(strsplit(gene, "_"), `[`, 1))

is_na <- df_beta$gene_symbol == "NA"
df_beta$gene_symbol[is_na] <- df_beta$gene[is_na]

ggplot(df_beta, aes(x = beta_msi, y = 1 / chi_msi, color = is_sig)) + 
  geom_point(alpha = 0.8) +
  # geom_point(data = dplyr::filter(df_beta, is_mmr), alpha = 0.6) +
  scale_color_brewer(palette = "Set1", name = "'Significant'") +
  geom_label_repel(data = dplyr::filter(df_beta, 1 / chi_msi > chi_cutoff), aes(label = gene_symbol)) +
  xlab(expression(beta)) +
  ylab(expression(paste("[", chi, "]"^-1)))
  
```

Which genes are sig?

```{r dt-sig}
dplyr::filter(df_beta, is_sig) %>% dplyr::arrange(desc(abs(beta_msi))) %>% 
  dplyr::select(gene, is_mmr, beta_msi, alpha_msi) %>% 
  datatable()
```

We can plot expression along phenotime for the 35 _significant_ genes:

```{r exprs-along-pst, fig.width = 12, fig.height = 9}
tmap <- pcavi$m_t


top_genes <- dplyr::filter(df_beta, is_sig) %>% 
  dplyr::arrange(desc(abs(beta_msi))) %>% 
  .$gene %>% head(n=20)

t(exprs(sce))[, top_genes] %>% 
  as_data_frame() %>% 
  dplyr::mutate(phenotime = tmap, x = factor(x)) %>% 
  gather(gene, expression, -phenotime, -x) %>% 
  ggplot(aes(x = phenotime, y = expression, color = x)) + geom_point(alpha = 0.4) +
  facet_wrap(~ gene, scales = "free_y") + 
  scale_color_brewer(palette = "Set1", name = "MSI") +
  theme(legend.position = "top", strip.text.x = element_text(size = 9)) +
  stat_smooth(se = FALSE, method = "lm", size = 2, formula = y ~ x) # ns(x,3))
```


What is phenotime? Check the immune response genes:

```{r immune-resp, fig.width = 7}
immune_genes <- c("FOXP3", "CD4", "IL2RB")
immune_inds <- match(immune_genes, fData(sce)$hgnc_symbol)

t(exprs(sce))[, immune_inds, drop=FALSE] %>% 
  as_data_frame() %>% 
  dplyr::mutate(phenotime = tmap, x = factor(x)) %>% 
  gather(gene, expression, -phenotime, -x) %>% 
  ggplot(aes(x = phenotime, y = expression, color = x)) + geom_point(alpha = 0.4) +
  facet_wrap(~ gene, scales = "free_y") + 
  scale_color_brewer(palette = "Set1", name = "MSI") +
  theme(legend.position = "top", strip.text.x = element_text(size = 9)) +
  stat_smooth(se = FALSE, method = "lm", size = 2)#, formula = y ~ ns(x,3))
```

Nice immune response plot:

```{r nice-immune-response}
df_immune <- t(exprs(sce))[, immune_inds, drop=FALSE] %>% 
  as_data_frame() %>% 
  dplyr::mutate(phenotime = tmap, x = factor(x)) %>% 
  gather(gene, expression, -phenotime, -x)

df_immune <- mutate(df_immune, 
                  gene_symbol = sapply(strsplit(gene, "_"), `[`, 1))

ggplot(df_immune, aes(x = phenotime, y = expression, fill = gene_symbol)) + 
  geom_point(shape = 21, color = 'grey10') +
  facet_wrap(~ gene_symbol, scales = "free_y", ncol = 1) + 
  scale_fill_brewer(palette = "Pastel1", name = "Gene") +
  #stat_smooth(se = FALSE, method = "lm", size = 1.5, 
  #            formula = y ~ ns(x,3), color = 'black') +
  ylab(expression(paste("Expression log"[2], "(TPM+1)"))) +
  xlab("Pathway score") +
  theme(legend.position = "none", 
        strip.text = element_text(size = 9),
        strip.background = element_rect(colour = "black", fill = "grey95", 
                                          size = 0.5, linetype = 1),
        axis.text = element_text(size = 9),
        axis.title = element_text(size = 10))

saveRDS(last_plot(), "../../figs/coad/tregs.rds")

```




Look at MLH1 in more detail:

```{r mlh1}
mlh1 <- c("MLH1")

make_gene_plot <- function(genes) {
  inds <- match(genes, fData(sce)$hgnc_symbol)
  
  df <- t(exprs(sce))[, inds, drop=FALSE] 
  colnames(df) <- fData(sce)$hgnc_symbol[inds]
  
  df <- as_data_frame(df) %>% 
    dplyr::mutate(phenotime = tmap, x = as.character(x)) %>% 
    gather(gene, expression, -phenotime, -x)
  
  df$x[is.na(df$x)] <- "NA"
  df$x <- plyr::mapvalues(df$x, from = sort(unique(df$x)), to = c("MSI-low", "MSI-high"))
  
  df$x <- factor(df$x, levels = c("MSI-low", "MSI-high"))
  
  df_nox <- dplyr::select(df, -x)
  
  ggplot(df, aes(x = phenotime, y = expression, fill = x, color = x)) + 
    geom_point(data = df_nox, color = "grey80", fill = "grey80") +
    geom_point(shape = 21, color = 'grey30', alpha = 0.7) +
    facet_grid(gene ~ x, scales = "free") + 
    scale_fill_brewer(palette = "Set1", name = "MSI") +
    scale_color_brewer(palette = "Set1", name = "MSI") +
    theme(legend.position = "none", strip.text = element_text(size = 9),
          strip.background = element_rect(colour = "black", fill = "grey95", 
                                          size = 0.5, linetype = 1)) +
  theme(axis.text = element_text(size = 9),
        axis.title = element_text(size = 10)) +
    stat_smooth(se = FALSE, method = "lm", size = 1.5, formula = y ~ ns(x,3),
                color = 'grey30') +
    ylab(expression(Expression ~ log[2] ~ "(TPM + 1)")) + 
    xlab("Pathway score (immune response)")
}

genes <- c("MLH1", "TGFBR2")

make_gene_plot(genes)

gene_plot <- last_plot()
```

# Differential expression comparison

We'll use limma-voom for standard differential expression as to whether a tumour is MSI-high or not:

```{r limma}
# retained_fnames <- rownames(sc_tumour[rowSums(counts(sc_tumour)) > 20, ])
dge <- DGEList(counts(sce))
dge <- calcNormFactors(dge)

sce$msi_cov <- 1 * (sce$msi_status == "msi-h")
design <- model.matrix(~ msi_cov, pData(sce))
v <- voom(dge, design, plot = TRUE)

fit <- lmFit(v, design)
fit <- eBayes(fit)
results <- decideTests(fit)
# vennDiagram(results)
```

And merge the results with phenotime:

```{r merge-with-us}
int_df_limma <- dplyr::rename(df_beta, feature = gene)

qvals <- p.adjust(fit$p.value[,2], method = 'BH')

df_limma <- data_frame(msi_coef = fit$coefficients[,2], 
                       msi_qval = qvals,
                       mu = pcavi$m_mu,
                       feature = featureNames(sce)) %>% 
  left_join(int_df_limma, by = "feature") %>% 
  dplyr::mutate(msi_log10qval = -log10(msi_qval),
                limma_sig = msi_qval < 0.05) 
```

Then we can compare in several ways. Firstly, are the $\alpha$ values consistent with the limma voom reported coefficients?

```{r compare-limma-alpha}
ggplot(df_limma, aes(x = msi_coef, y = alpha_msi)) + 
  geom_point(alpha = 0.5) +
  xlab("Limma voom coefficient") + ylab(expression(paste("Phenotime ", alpha)))

ggplot(df_limma, aes(x = msi_coef, y = alpha_msi, color = beta_msi)) + 
  geom_point(alpha = 0.5) +
  xlab("Limma voom coefficient") + 
  ylab(expression(paste("Phenotime ", alpha))) +
  scale_color_viridis()
```

There is some "twisting" there which may be due to using a mixed effects model rather than straight forward DE.

Is there any relationship between the limma voom reported coefficients and phenotimes $\beta$ values?

```{r compare-beta}

cols <- RColorBrewer::brewer.pal(3, "Set2")
cols2 <- c("#c5e2d9", cols[2])

ggplot(df_limma, x =-log10(msi_qval), aes(x = beta_msi, y =-log10(msi_qval), color = is_sig_graph)) + 
  geom_point() +
  ylab(expression(paste("Limma voom -", log[10], "(q-value)"))) + 
  xlab(expression(paste("PhenoPath ", beta))) +
  # theme(legend.position = 'top') +
  geom_hline(yintercept = -log10(0.05), linetype = 2, alpha = 0.5) +
  theme(axis.text = element_text(size = 8),
        axis.title = element_text(size = 9)) +
  #theme(legend.title = element_text(size = 10),
  #      legend.text = element_text(size = 9)) +
  theme(legend.position = "none") +
  scale_color_manual(values = cols2, name = "Interaction") 

limma_plot <- ggExtra::ggMarginal(last_plot(), margins = "y", type = "histogram", size = 10)
```

So no obvious correlations. Let's just double check the $p$-values:

```{r limma-pval}
ggplot(df_limma, aes(x = beta_msi, y = msi_log10qval)) + geom_point(alpha = 0.5) +
  ylab("Limma voom -log10 pval") + xlab(expression(paste("Phenotime ", beta)))
```

Again no obvious relationship.

# Crossover bit for chris

```{r df-beta}
df_beta <- mutate(df_beta,
                  crossover = - alpha_msi / beta_msi)

filter(df_beta, is_sig) %>% 
  ggplot(aes(x = crossover)) + geom_histogram()
```



```{r more-vis}
textinfo <- frame_data(
  ~x, ~y, ~label,
  0.5, 0.1, "Gene upregulated\nMSI increases upregulation",
  0.5, -0.1, "Gene upregulated\nMSI decreases upregulation",
  -0.6, 0.1, "Gene downregulated\nMSI decreases downregulation",
  -0.6, -0.1, "Gene downregulated\nMSI increases downregulation"
)

cols <- RColorBrewer::brewer.pal(3, "Set2")
cols2 <- c("#c5e2d9", cols[2])

ggplot(df_beta, aes(x = c, y = beta_msi, color = is_sig_graph)) + 
  geom_point() +
  geom_vline(xintercept = 0, linetype = 2, alpha = 0.5) +
  geom_hline(yintercept = 0, linetype = 2, alpha = 0.5) +
  scale_color_manual(values = cols2, name = "Interaction") +
  geom_text_repel(data = dplyr::filter(df_beta, is_sig, beta_msi < -0.3 | beta_msi > 0.38),
                  aes(label = hgnc_symbol), color = 'black', 
                  size = 2.5) +
  ylab("Covariate pathway interaction") +
  xlab("Pathway loading") +
  theme(legend.position = 'bottom',
        axis.text = element_text(size = 10),
        axis.title = element_text(size = 11),
        legend.title = element_text(size = 11),
        legend.text = element_text(size = 10)) +
  geom_text(data = textinfo, aes(x = x, y = y, label = label), 
            color = 'black', size = 3)

lplot <- last_plot()
lplot
```

# What is pseudotime (2)


```{r get-correlations}
library(goseq)
genome <- "hg19"
id <- "ensGene"

all_genes <- fData(sce)$ensembl_gene_id

cs <- apply(exprs(sce), 1, cor, tmap)
cdf <- data_frame(feature = all_genes, correlation = cs)

upreg_genes <- filter(cdf, correlation > 0.5) %>% extract2("feature") 
downreg_genes <- filter(cdf, correlation < -0.5) %>% extract2("feature")

upg <- 1 * (all_genes %in% upreg_genes)
downg <- 1 * (all_genes %in% downreg_genes)

names(upg) <- names(downg) <- all_genes

pwfup <- nullp(upg, genome, id)
goup <- goseq(pwfup, genome, id, test.cats = "GO:BP")

pwfdown <- nullp(downg, genome, id)
godown <- goseq(pwfdown, genome, id, test.cats = "GO:BP")
```

Graph results:

```{r graph-res}
parse_go <- function(go, type, n_tested) {
  go <- go %>% 
    mutate(qval = p.adjust(over_represented_pvalue)) %>% 
    mutate(log10qval = -log10(qval),
           prop_in_cat = numInCat / n_tested) %>% 
    head(n = 12) %>% 
    mutate(type = type) %>% 
    tbl_df() %>% 
    arrange(desc(log10qval))
  go
}

gos <- bind_rows(
  parse_go(goup, "Up-regulated", length(upreg_genes)),
  parse_go(godown, "Down-regulated", length(downreg_genes))
)

gos$term <- factor(gos$term, levels = gos$term[order(gos$log10qval)])

gos <- mutate(gos, prop_cat = numDEInCat / numInCat)

filter(gos, type == "Up-regulated") %>% 
  ggplot(aes(x = term, y = log10qval)) + #, size = 100 * prop_cat)) +
    geom_point() +
    coord_flip() +
    theme(axis.title.y = element_blank()) +
    ylab(expression(paste("-", log[10], " q-value"))) +
  scale_size(name = "% category\nrepresented") +
    theme(#legend.position = c(-.2, -.14), legend.direction = "horizontal",
      legend.position = c(0, 0), legend.direction = "horizontal",
        axis.text = element_text(size = 8),
        axis.title = element_text(size = 10),
        legend.title = element_text(size = 11),
        legend.text = element_text(size = 10)) 

goplot <- last_plot()
```

Save df_beta:

```{r save-df-beta}
filter(df_beta, is_sig) %>% 
  write_csv("../../data/COAD/coad_interactions.csv")
```


```{r large-plot}
lsize <- 11
bottom_left <- plot_grid(plot_grid(NULL, limma_plot, NULL, nrow = 1, rel_widths = c(1,6, 1)), 
                         goplot,
                         ncol = 1, labels = c("B", "C"),
                         label_size = lsize)
bottom_grid <- plot_grid(bottom_left, gene_plot, nrow = 1, labels = c("", "D"),
                         label_size = lsize)

p <- plot_grid(lplot, bottom_grid, ncol = 1, 
               rel_heights = c(2,1.2), labels = c("A",""),
          label_size = lsize)

print( p )

saveRDS(limma_plot, "../../figs/coad/limma_plot.rds")
saveRDS(goplot, "../../figs/coad/goplot.rds")
saveRDS(gene_plot, "../../figs/coad/gene_plot.rds")
saveRDS(lplot, "../../figs/coad/lplot.rds")
```

# Cibertsort

```{r csort}
csort_raw <- read_tsv("../../data/COAD/cellTypeFractionsAll.tsv")

csort <- dplyr::mutate(csort_raw, patient_barcode = tolower(patientBarcode)) %>% 
  dplyr::select(-patientBarcode, -fair_LM22)

csort_wide <- spread(csort, cell_type, cibersort_LM22)

sample_names <- sce$patient_barcode
mm <- match(sample_names, csort_wide$patient_barcode)
csort <- csort_wide[mm, ]
stopifnot(all.equal(sce$patient_barcode, csort$patient_barcode))

tmap <- pcavi$m_t

cors <- apply(as.matrix(csort[,-c(1:2)]), 2, cor, tmap, method = "spearman")

cdf <- data_frame(cell_type = names(csort)[-c(1:2)],
                  correlation_to_z = cors)
cdf <- dplyr::arrange(cdf, correlation_to_z)
cell_type_levels <- cdf$cell_type
cdf$cell_type <- factor(cdf$cell_type, levels = cell_type_levels)

ggplot(cdf, aes(x = cell_type, y = correlation_to_z)) +
  geom_bar(stat = "identity") +
  ylab("Cibertsort correlation to z") +
  xlab("Cibertsort cell type") +
  coord_flip()
```

Boxplot for Chris:

```{r}
genes <- c("MLH1", "TGFBR2")

mm <- match(genes, fData(sce)$hgnc_symbol)

df_bp <- as_data_frame(t(exprs(sce)[mm, ]))
names(df_bp) <- genes
df_bp_tidy <- mutate(df_bp, MSI_high = sce$msi_status == "msi-h") %>% 
  gather(gene, expression, -MSI_high)

filter(df_bp_tidy) %>% 
  ggplot(aes(x = gene, y = expression, fill = MSI_high)) +
    geom_violin() +
    scale_fill_brewer(palette = "Set1", name = "High MSI") +
    labs(x = "Gene", y = "Normalised Expression")


ggsave("../../figs/coad_violinplot.png", width = 5, height = 4)
```
