---
title: "BRCA using Variational Bayes Phenotime"
author: "Kieran Campbell"
date: "December 2017"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = FALSE, fig.width = 6, fig.height = 4)

library(scater)
library(tidyverse)
library(magrittr)
library(matrixStats)
library(ggrepel)
library(cowplot)
library(viridis)
library(splines)
library(DT)
library(limma)
library(edgeR)
library(RColorBrewer)
# library(reactome.db)
library(phenopath)
library(biomaRt)
library(goseq)

set.seed(123L)

scale_vec <- function(x) (x - mean(x)) / sd(x)

select <- dplyr::select
filter <- dplyr::filter
mutate <- dplyr::mutate
arrange <- dplyr::arrange

lsize = 12
```

# Load data

Load results:

```{r load-all-data}
pcavi <- readRDS("../../data/BRCA/clvm_results.rds")
sce <- readRDS("../../data/BRCA/sce_brca_clvm.rds")
sce <- updateSCESet(sce)
x <- sce$x
sce$tmap <- pcavi$m_z

```




Plot the elbo:

```{r plot-elbo}
plot_elbo(pcavi) +
  xlab(expression("CAVI iterations"~x10^{-1})) +
  theme_cowplot(font_size = 11)

ggsave("../../figs/elbo_brca.png", width = 5, height = 3)
```

# Exploring phenotime results

## Comparison to PC1

```{r pseudotime-vs-pc1}
pc1 <- prcomp(t(exprs(sce)), scale = TRUE)$x[,1]

qplot(scale_vec(pc1), scale_vec(pcavi$m_z), color = factor(x)) +
  scale_color_brewer(palette = "Set1", name = "x") +
  stat_function(fun = function(x) x, color = 'black') + 
  xlab("PC1") + ylab(expression(z))
```

## Identifying significant genes

In the context of CAVI we have modelled each interaction coefficient by an approximating distribution of the form

$$ \beta_{pg} \sim \text{Normal}(m_{\beta_{pg}}, s_{\beta_{pg}}^2) $$
We define a _significant_ interaction if the $2 \sigma$ posterior doesn't cross zero, ie

$$ m_{\beta_{pg}} - n s_{\beta_{pg}} > 0 $$
or
$$ m_{\beta_{pg}} + n s_{\beta_{pg}} < 0 $$

Here we use $n=3$:

```{r beta-df}
df_beta <- data_frame(
  c = pcavi$m_lambda,
  beta = pcavi$m_beta[1,],
  chi = pcavi$chi_exp[1,],
  alpha = pcavi$m_alpha[1,],
  pos_sd = sqrt(pcavi$s_beta[1,]),
  gene = rownames(sce),
  is_sig = as.vector(significant_interactions(pcavi, n = 3)),
  hgnc_symbol = rowData(sce)$hgnc_symbol,
  ensembl_gene_id = rowData(sce)$ensembl_gene_id
)

write_csv(df_beta, "../../data/BRCA/df_gene_brca.csv")

df_beta$is_sig_graph <- 
  plyr::mapvalues(df_beta$is_sig, from = c(FALSE, TRUE),
                  to = c("Non-significant", "Significant"))
```

This gives:

```{r how-many-sig}
print(sum(df_beta$is_sig))
mean(df_beta$is_sig)
```

_significant_ interactions.

We can plot these:

```{r plot-sigs}
chi_cutoff <- 0.125
df_beta <- mutate(df_beta, 
                  gene_symbol = sapply(strsplit(gene, "_"), `[`, 1))

is_na <- df_beta$gene_symbol == "NA"
df_beta$gene_symbol[is_na] <- df_beta$gene[is_na]

ggplot(df_beta, aes(x = beta, y = 1 / chi, color = is_sig)) + 
  geom_point(alpha = 0.8) +
  # geom_point(data = dplyr::filter(df_beta, is_mmr), alpha = 0.6) +
  scale_color_brewer(palette = "Set1", name = "'Significant'") +
  geom_label_repel(data = dplyr::filter(df_beta, 1 / chi > chi_cutoff), aes(label = gene_symbol)) +
  xlab(expression(beta)) +
  ylab(expression(paste("[", chi, "]"^-1)))
  
```

Which genes are sig?

```{r dt-sig}
filter(df_beta, is_sig) %>% arrange(desc(abs(beta))) %>% 
  select(gene, beta, alpha) %>% 
  datatable()
```

We can plot expression along phenotime for the 35 _significant_ genes:

```{r exprs-along-pst, fig.width = 12, fig.height = 9}
tmap <- pcavi$m_z
sce$tmap <- tmap


top_genes <- dplyr::filter(df_beta, is_sig) %>% 
  dplyr::arrange(desc(abs(beta))) %>% 
  extract2("gene") %>% head(n=20)

t(exprs(sce))[, top_genes] %>% 
  as_data_frame() %>% 
  dplyr::mutate(phenotime = tmap, x = factor(x)) %>% 
  gather(gene, expression, -phenotime, -x) %>% 
  ggplot(aes(x = phenotime, y = expression, color = x)) + geom_point(alpha = 0.4) +
  facet_wrap(~ gene, scales = "free_y") + 
  theme(legend.position = "top", strip.text.x = element_text(size = 9)) +
  stat_smooth(se = FALSE, method = "lm", size = 2) + 
  scale_color_brewer(name = "ER_status", palette = 'Set1')
```

# Plot top genes

```{r top-genes, fig.width = 7, fig.height=12}
top_genes <- dplyr::filter(df_beta, is_sig) %>% 
  arrange(desc(abs(beta))) %>% 
  extract2("gene")

top_genes <- top_genes[1:4]

plot_nice_expression <- function(gene) {
  df_gex <- t(exprs(sce))[, gene, drop=FALSE] %>% 
    as_data_frame() %>% 
    dplyr::mutate(phenotime = tmap, x = as.character(x)) %>% 
    gather(gene, expression, -phenotime, -x)
  
  df_gex$x[is.na(df_gex$x)] <- "NA"
  df_gex$x <- plyr::mapvalues(df_gex$x, from = sort(unique(df_gex$x)), to = c("ER negative", "ER positive"))
  
  df_gex$x <- factor(df_gex$x, levels = c("ER negative", "ER positive"))
  
  df_gex_nox <- dplyr::select(df_gex, -x)
  
  ggplot(df_gex, aes(x = phenotime, y = expression, color = x)) + 
    geom_point(data = df_gex_nox, color = "grey70") +
    geom_point() +
    facet_wrap(~ x, scales = "free_y") + 
    scale_color_brewer(palette = "Set1", name = "ER status") +
    theme(legend.position = "none", strip.text.x = element_text(size = 9),
          axis.title = element_text(size = 10),
          axis.text = element_text(size = 9),
          plot.title = element_text(size = 11)) +
    stat_smooth(se = FALSE, method = "lm", size = 1.5, color = 'black') + # formula = y ~ ns(x,3), 
    ggtitle(strsplit(gene, "_")[[1]][1]) +
    ylab(expression(paste("Expression log"[2], "(TPM+1)"))) +
    xlab("z")
}

plots <- lapply(top_genes, plot_nice_expression)

plot_grid(plotlist = plots, ncol = 1)
```

# Differential expression comparison


```{r limma}
# retained_fnames <- rownames(sc_tumour[rowSums(counts(sc_tumour)) > 20, ])
dge <- DGEList(counts(sce))
dge <- calcNormFactors(dge)

design <- model.matrix(~ x, colData(sce))
v <- voom(dge, design, plot = TRUE)

fit <- lmFit(v, design)
fit <- eBayes(fit)
results <- decideTests(fit)
# vennDiagram(results)
```

And merge the results with phenotime:

```{r merge-with-us}
int_df_limma <- dplyr::rename(df_beta, feature = gene)

qvals <- p.adjust(fit$p.value[,2], method = 'BH')

df_limma <- data_frame(coef = fit$coefficients[,2], 
                       pval = fit$p.value[,2],
                       qval = qvals,
                       # mu = pcavi$m_mu,
                       feature = rownames(sce)) %>% 
  left_join(int_df_limma, by = "feature") %>% 
  dplyr::mutate(log10qval = -log10(qval),
                limma_sig = qval < 0.05) 
```

Then we can compare in several ways. Firstly, are the $\alpha$ values consistent with the limma voom reported coefficients?

```{r compare-limma-alpha}
# ggplot(df_limma, aes(x = coef, y = alpha)) + 
#   geom_point(alpha = 0.5) +
#   xlab("Limma voom coefficient") + ylab(expression(paste("Phenotime ", alpha)))
# 
# ggplot(df_limma, aes(x = coef, y = alpha, color = mu)) + 
#   geom_point(alpha = 0.5) +
#   xlab("Limma voom coefficient") + 
#   ylab(expression(paste("Phenotime ", alpha))) +
#   scale_color_viridis(name = expression(mu))
```

There is some "twisting" there which may be due to using a mixed effects model rather than straight forward DE.

Nice plot for paper:


```{r compare-beta}

cols <- RColorBrewer::brewer.pal(3, "Set2")
cols2 <- c("#c5e2d9", cols[2])

ggplot(df_limma, x =-log10(qval), aes(x = beta, y =-log10(qval), color = is_sig_graph)) + 
  geom_point() +
  ylab(expression(paste("Limma voom -", log[10], "(q-value)"))) + 
  xlab(expression(paste("PhenoPath ", beta))) +
  # theme(legend.position = 'top') +
  geom_hline(yintercept = -log10(0.05), linetype = 2, alpha = 0.5) +
  theme(axis.text = element_text(size = 9),
        axis.title = element_text(size = 10)) +
  #theme(legend.title = element_text(size = 10),
  #      legend.text = element_text(size = 9)) +
  theme(legend.position = "none") +
  scale_color_manual(values = cols2, name = "Interaction") 

limma_plot <- ggExtra::ggMarginal(last_plot(), margins = "y", type = "histogram", size = 10)
```

# What is pseudotime?

```{r get-correlations}
library(goseq)
genome <- "hg19"
id <- "ensGene"

all_genes <- rowData(sce)$ensembl_gene_id

cs <- apply(exprs(sce), 1, cor, tmap)
cdf <- data_frame(feature = all_genes, correlation = cs)

upreg_genes <- filter(cdf, correlation > 0.5) %>% extract2("feature") 
downreg_genes <- filter(cdf, correlation < -0.5) %>% extract2("feature")

upg <- 1 * (all_genes %in% upreg_genes)
downg <- 1 * (all_genes %in% downreg_genes)

names(upg) <- names(downg) <- all_genes

pwfup <- nullp(upg, genome, id)
goup <- goseq(pwfup, genome, id, test.cats = "GO:BP")

pwfdown <- nullp(downg, genome, id)
godown <- goseq(pwfdown, genome, id, test.cats = "GO:BP")
```

Graph results:

```{r graph-res}
parse_go <- function(go, type, n_tested) {
  go <- go %>% 
    mutate(qval = p.adjust(over_represented_pvalue)) %>% 
    mutate(log10qval = -log10(qval),
           prop_in_cat = numInCat / n_tested) %>% 
    head(n = 12) %>% 
    mutate(type = type) %>% 
    tbl_df() %>% 
    arrange(desc(log10qval))
  go
}

gos <- bind_rows(
  parse_go(goup, "Up-regulated", length(upreg_genes)),
  parse_go(godown, "Down-regulated", length(downreg_genes))
)

gos$term <- stringr::str_to_title(gos$term)
gos$term <- factor(gos$term, levels = gos$term[order(gos$log10qval)])

filter(gos, type == "Up-regulated") %>% 
  ggplot(aes(x = term, y = log10qval)) +
    geom_point() +
    # facet_wrap(~ type, scales = "free_y", nrow = 2) +
    coord_flip() +
    theme(axis.title.y = element_blank(), legend.position = "none") +
    # scale_color_brewer(palette = "Set1") +
    ylab(expression(paste("-", log[10], " q-value"))) +
    theme(#legend.position = c(-.2, -.14), legend.direction = "horizontal",
      legend.position = c(0, 0), legend.direction = "horizontal",
        axis.text = element_text(size = 8),
        axis.title = element_text(size = 10),
        legend.title = element_text(size = 11),
        legend.text = element_text(size = 10)) +
    geom_segment(aes(y = min(log10qval - 1), yend = log10qval, x = term, xend = term),
               color = 'grey30', linetype = 3)

goplot <- last_plot()

```



Do the same for beta:

```{r go-beta}
genome <- "hg19"
id <- "ensGene"

all_genes <- rowData(sce)$ensembl_gene_id
df_beta$ensembl_gene_id <- all_genes


upreg_genes <- filter(df_beta, beta > .5 & is_sig) %>% extract2("ensembl_gene_id")
downreg_genes <- filter(df_beta, beta < -.5 & is_sig) %>% extract2("ensembl_gene_id")

# upreg_genes <- c(upreg_genes, downreg_genes)

upg <- 1 * (all_genes %in% upreg_genes)
downg <- 1 * (all_genes %in% downreg_genes)

names(upg) <- names(downg) <- all_genes

pwfup <- nullp(upg, genome, id)
goup <- goseq(pwfup, genome, id, test.cats = "GO:BP")

pwfdown <- nullp(downg, genome, id)
godown <- goseq(pwfdown, genome, id, test.cats = "GO:BP")
```

Graph results:

```{r parse-go}
parse_go <- function(go, type, n_tested) {
  go <- go %>% 
    mutate(qval = p.adjust(over_represented_pvalue)) %>% 
    mutate(log10qval = -log10(qval),
           prop_in_cat = numInCat / n_tested) %>% 
    head(n = 10) %>% 
    mutate(type = type) %>% 
    tbl_df() %>% 
    arrange(desc(log10qval))
  go
}

gos <- bind_rows(
  parse_go(goup, "Up-regulated", length(upreg_genes)),
  parse_go(godown, "Down-regulated", length(downreg_genes))
)

gos$term <- factor(gos$term, levels = gos$term[order(gos$log10qval)])


ggplot(gos, aes(x = term, y = log10qval, color = type)) +
  geom_point() +
  facet_wrap(~ type, scales = "free_y", nrow = 2) +
  coord_flip() +
  theme(axis.title.y = element_blank(), legend.position = "none") +
  scale_color_brewer(palette = "Set1") +
  ylab(expression(paste(log[10], " q-value"))) +
  theme(axis.text = element_text(size = 10),
        axis.title.x = element_text(size = 11))
```

Compare to known coefficients:

```{r compare-known-coef, eval = FALSE}
df <- dplyr::select(colData(sce), z = tmap, Tumor, Node, Metastasis, AJCC_Stage, PAM50_mRNA, RPPA_Clusters)

df <- gather(df, variable, value, -z)

variables <- unique(df$variable)

plots <- lapply(variables, function(v) {
  filter(df, variable == v) %>% 
    ggplot(aes(x = value, y = z)) + geom_boxplot() +
    ggtitle(v) + 
    theme(axis.text.x = element_text(size = 9, angle = 90, hjust = 1),
          axis.title.x = element_blank())
})

plot_grid(plotlist = plots)
```

```{r basal, eval = FALSE}
gene <- "FOXC1_ENSG00000054598"
gene <- "FBP1_ENSG00000165140"

is_basal <- sce$PAM50_mRNA == "Basal-like"

  df_gex <- t(exprs(sce))[, gene, drop=FALSE] %>% 
    as_data_frame() %>% 
    dplyr::mutate(phenotime = tmap, x = is_basal) %>% 
    gather(gene, expression, -phenotime, -x)
  
  df_gex$x[is.na(df_gex$x)] <- "NA"
  df_gex$x <- plyr::mapvalues(df_gex$x, from = sort(unique(df_gex$x)), 
                              to = c("Not basal", "NA", "Basal"))
  
  df_gex$x <- factor(df_gex$x, levels = c("Not basal", "NA", "Basal"))
  df_gex <- filter(df_gex, x != "NA")
  
  df_gex_nox <- dplyr::select(df_gex, -x)
  
  ggplot(df_gex, aes(x = phenotime, y = expression, color = x)) + 
    geom_point(data = df_gex_nox, color = "grey70") +
    geom_point() +
    facet_wrap(~ x, scales = "free_y") + 
    scale_color_brewer(palette = "Set1", name = "ER status") +
    theme(legend.position = "none", strip.text.x = element_text(size = 9),
          axis.title = element_text(size = 10),
          axis.text = element_text(size = 9),
          plot.title = element_text(size = 11)) +
    stat_smooth(se = FALSE, method = "lm", size = 1.5, color = 'black') + # formula = y ~ ns(x,3), 
    ggtitle(strsplit(gene, "_")[[1]][1]) +
    ylab(expression(paste("Expression log"[2], "(TPM+1)"))) +
    xlab("z")

```

Time for some reactome fun: NOT EVAL'D

```{r reactome-time, eval = FALSE}
pathways <- c("1643713", "1226099",
              "2219528", "2644603",
              "3304351", "4791275")

id_to_name <- as.list(reactomePATHID2NAME)
pathway_names <- id_to_name[pathways]

pathways_to_genes <- as.list(reactomePATHID2EXTID)

gene_list <- pathways_to_genes[pathways]
pathway_names <- sapply(pathway_names, function(pn) {
  gsub("Homo sapiens: ", "", pn)
})
names(gene_list) <- pathway_names

mart <- useMart("ensembl", "hsapiens_gene_ensembl")

to_ensembl <- function(gl) {
  bm <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"),
      filters = c("entrezgene"),
      values = as.numeric(gl),
      mart = mart)
  return(bm$ensembl_gene_id)
}

gene_list_ensembl <- lapply(gene_list, to_ensembl)
```

And graph the results for various metrics:

```{r graph-for-metrics, eval = FALSE, eval = FALSE}
sce <- readRDS("../../data/BRCA/sce_brca_gene_level.rds")

all_genes <- unique(unlist(gene_list_ensembl))
all_genes <- all_genes[all_genes %in% rowData(sce)$ensembl_gene_id]
mm <- match(all_genes, rowData(sce)$ensembl_gene_id)
is_basal <- sce$PAM50_mRNA == "Basal-like"

Y <- t(exprs(sce)[mm, ])
colnames(Y) <- all_genes

df_gex <- Y %>% 
  as_data_frame() %>% 
  dplyr::mutate(is_basal, z = colData(sce)[['tmap']]) %>% 
  gather(gene, expression, -is_basal, -z)

df_pheno <- bind_rows(
  lapply(names(gene_list_ensembl), function(n) {
    data_frame(pathway = n, gene = gene_list_ensembl[[n]])
  })
)

df <- inner_join(df_gex, df_pheno, by = 'gene')

df2 <- filter(df, !is.na(is_basal)) %>% 
  group_by(gene, pathway, is_basal) %>% 
  summarise(gex = mean(expression))

df3 <- df %>% # filter(df, !is.na(is_basal)) %>% 
  group_by(pathway, z) %>% 
  summarise(gex = median(expression))

ggplot(df3, aes(x = z, y = gex)) + 
  geom_point(alpha = 0.7) + 
  facet_wrap(~ pathway, scales = 'free_y') +
  ylab("Median pathway expression") +
  stat_smooth(color = 'red', se = F)


df4 <- df %>% # filter(df, !is.na(is_basal)) %>% 
  group_by(pathway, z) %>% 
  summarise(gex = var(expression))

# ggplot(df4, aes(x = z, y = gex)) + 
#   geom_point(alpha = 0.7) + 
#   facet_wrap(~ pathway, scales = 'free_y') +
#   ylab("Variance in pathway expression") +
#   stat_smooth(color = 'red', se = F)
```





# Crossover bit for chris

```{r df-beta}
df_beta <- mutate(df_beta,
                  crossover = - alpha / beta)

filter(df_beta, is_sig) %>% 
  ggplot(aes(x = crossover)) + 
  geom_histogram(fill = "#74a9cf", color = "grey90", bins = 30) +
  xlab("Crossover point") + ylab("Number of genes") +
  theme(axis.text = element_text(size = 9),
        axis.title = element_text(size = 10))

cross_plot <- last_plot()

ggsave("../../figs/supplementary_crossover.png", width = 6, height = 5)
```

Crossover GO analysis:

```{r crossover-go}
library(goseq)
genome <- "hg19"
id <- "ensGene"

all_genes <- rowData(sce)$ensembl_gene_id

crossover_genes <- filter(df_beta, 
                          is_sig & crossover > 0.4) %>% 
  extract2("ensembl_gene_id") 

crossg <- 1 * (all_genes %in% crossover_genes)
names(crossg) <- all_genes

pwfcross <- nullp(crossg, genome, id)
gocross <- goseq(pwfcross, genome, id, test.cats = "GO:BP")

pgo <- parse_go(gocross, "crossover", length(all_genes))
```

Crossover plots:

```{r crossover-gene-plots}
tmap <- pcavi$m_z

cross_df <- dplyr::select(df_beta, gene, crossover, hgnc_symbol)

top_genes <- dplyr::filter(df_beta, is_sig) %>% 
  arrange(desc(abs(beta))) %>% 
  extract2("gene") %>% head(n=12)

df_gex <- t(exprs(sce))[, top_genes, drop=FALSE] %>% 
  as_data_frame() %>% 
  dplyr::mutate(phenotime = tmap, x = as.character(x)) %>% 
  gather(gene, expression, -phenotime, -x)

df_gex$x[is.na(df_gex$x)] <- "NA"
df_gex$x <- plyr::mapvalues(df_gex$x, from = sort(unique(df_gex$x)), to = c("ER negative", "ER positive"))

df_gex$x <- factor(df_gex$x, levels = c("ER negative", "ER positive"))

df_gex <- inner_join(df_gex, cross_df, by = "gene")

ggplot(df_gex, aes(x = phenotime, y = expression, color = x)) + 
  geom_point(alpha = 0.1) +
  facet_wrap(~ hgnc_symbol, scales = "free_y", ncol = 3) + 
  theme(legend.position = "top", strip.text.x = element_text(size = 9)) +
  stat_smooth(se = FALSE, method = "lm", size = 2) + 
  scale_color_brewer(name = "", palette = 'Set1') +
  geom_vline(aes(xintercept = crossover), linetype = 2) +
  xlab("Pathway score") +
  ylab(expression(Expression ~ log[2] ~ "(TPM + 1)")) 

# saveRDS(last_plot(), "../../figs/brca/crossover_thesis.rds")
# 
# ggsave("../../figs/supplementary_crossover_2.png", width = 10, height = 9)
```

Save df_beta:

```{r save-df-beta}
filter(df_beta, is_sig) %>% 
  write_csv("../../data/BRCA/brca_interactions.csv")
```

# Plots for paper

```{r more-vis}
textinfo <- frame_data(
  ~x, ~y, ~label,
  0.6, 0.15, "Gene upregulated\nER+ increases upregulation",
  0.6, -0.15, "Gene upregulated\nER+ decreases upregulation",
  -0.7, 0.15, "Gene downregulated\nER+ decreases downregulation",
  -0.7, -0.15, "Gene downregulated\nER+ increases downregulation"
)

cols <- RColorBrewer::brewer.pal(3, "Set2")
cols2 <- c("#c5e2d9", cols[2])


outline_cols = c("#c5e2d9", 'black')

ggplot(df_beta, aes(x = c, y = beta)) + 
  geom_point(shape = 21, aes(fill = is_sig_graph, color = is_sig_graph), alpha = 0.8) +
  geom_vline(xintercept = 0, linetype = 2, alpha = 0.5) +
  geom_hline(yintercept = 0, linetype = 2, alpha = 0.5) +
  scale_fill_manual(values = cols2, name = "Interaction") +
  scale_color_manual(values = outline_cols, name = "Interaction") +
  geom_text_repel(data = dplyr::filter(df_beta, is_sig, abs(beta) > 0.7),
                 aes(label = hgnc_symbol), color = 'black',
                 size = 3) +
  ylab("Covariate-pseudotime interaction") +
  xlab("Gene regulation over pseudotime") +
  theme(legend.position = 'bottom',
        axis.text = element_text(size = 10),
        axis.title = element_text(size = 11),
        legend.title = element_text(size = 11),
        legend.text = element_text(size = 10)) +
  geom_text(data = textinfo, aes(x = x, y = y, label = label), 
            color = 'black', size = 3, fontface = "bold")

lplot <- last_plot()
lplot

```

```{r make-gene-plot}
tmap <- sce$tmap
make_gene_plot <- function(genes) {
  inds <- match(genes, rowData(sce)$hgnc_symbol)
  
  df <- t(exprs(sce))[, inds, drop=FALSE] 
  colnames(df) <- rowData(sce)$hgnc_symbol[inds]
  
  df <- as_data_frame(df) %>% 
    dplyr::mutate(phenotime = tmap, x = as.character(x)) %>% 
    gather(gene, expression, -phenotime, -x)
  
  cross_df_2 <- dplyr::select(cross_df, crossover, hgnc_symbol) %>% 
    dplyr::rename(gene = hgnc_symbol)
  df <- inner_join(df, cross_df_2, by = "gene")
  
  df$x[is.na(df$x)] <- "NA"
  df$x <- plyr::mapvalues(df$x, from = sort(unique(df$x)), to = c("ER-", "ER+"))
  
  df$x <- factor(df$x, levels = c("ER-", "ER+"))
  
  df_nox <- dplyr::select(df, -x)
  
  cols <- brewer.pal(5, "Set1")
  
  ggplot(df, aes(x = phenotime, y = expression, fill = x, color = x)) + 
    geom_point(data = df_nox, color = "grey80", fill = "grey80") +
    geom_point(shape = 21, color = 'grey30', alpha = 0.7) +
    facet_grid(x ~ gene, scales = "free") + 
    scale_fill_brewer(palette = "Set1", name = "MSI") +
    scale_color_brewer(palette = "Set1", name = "MSI") +
    theme(legend.position = "none", 
          strip.background = element_rect(colour = "black", fill = "grey95", 
                                          size = 0.5, linetype = 1),
          strip.text = element_text(face = "bold", size = 9),
          strip.text.x = element_text(margin = margin(.1, 0, .1, 0, unit = "cm")),
          strip.text.y = element_text(margin = margin(0, .1, 0, .1, unit = "cm"))) +
  theme(axis.text = element_text(size = 9),
        axis.title = element_text(size = 10)) +
    stat_smooth(se = FALSE, method = "lm", size = 1.5, #formula = y ~ ns(x,3),
                color = 'grey30') +
    ylab(expression(Expression ~ log[2] ~ "(TPM + 1)")) + 
    xlab("Pseudotime") +
    geom_vline(aes(xintercept = crossover), linetype = 2, color = "grey40")
}

genes <- c("ESR1", "FBP1", "FOXC1")

make_gene_plot(genes)

gene_plot <- last_plot()
```

New plot # 2:

```{r}
bottom_left_grid <- plot_grid(limma_plot, cross_plot, ncol = 1,
                              labels = c("b", "c"), label_size = lsize)
bottom_grid <- plot_grid(bottom_left_grid, gene_plot, nrow = 1,
                         labels = c("", "d"), label_size = lsize, rel_widths = c(1, 2))

brca_2 <- plot_grid(lplot, bottom_grid, ncol = 1, labels = c("a", ""), label_size = lsize,
                    rel_heights = c(4,3))

ggsave("../../figs/new/brca_2.png", width = 10, height = 9, dpi = 600)

ggsave("../../figs/new/brca_2.svg", width = 10, height = 9)

```

Old large plot:

```{r large-plot}
# lsize <- 11
# bottom_left <- plot_grid(plot_grid(NULL, limma_plot, NULL, nrow = 1, rel_widths = c(1,6, 1)), 
#                          goplot,
#                          ncol = 1, labels = c("F", "G"),
#                          label_size = lsize)
# bottom_grid <- plot_grid(bottom_left, gene_plot, nrow = 1, labels = c("", "H"),
#                          label_size = lsize)
# 
# plot_grid(lplot, bottom_grid, ncol = 1, rel_heights = c(2,1.2), labels = c("E",""),
#           label_size = lsize)
# 
# saveRDS(cross_plot, "../../figs/brca/cross_plot.rds")
# saveRDS(limma_plot, "../../figs/brca/limma_plot.rds")
# saveRDS(goplot, "../../figs/brca/goplot.rds")
# saveRDS(gene_plot, "../../figs/brca/gene_plot.rds")
# saveRDS(lplot, "../../figs/brca/lplot.rds")
```

New heatmap plots:

```{r, cache = FALSE}
sce_gene <- readRDS("../../data/BRCA/sce_brca_gene_level.rds")
sce_gene <- updateSCESet(sce_gene)
sce_gene <- sce_gene[, colnames(sce)]
angiogenesis_genes_all <- read_csv("../../data/BRCA/ag_hallmark_geneset.txt", skip = 2, col_names = FALSE)[[1]]
df_beta <- mutate(df_beta, is_angiogenesis = hgnc_symbol %in% angiogenesis_genes_all)

growth_factor_grep <- c("^FGF", "^VEGF", "^EGF", "^TGF")
growth_factors <- unlist(lapply(growth_factor_grep, grep, rowData(sce)$hgnc_symbol, value = TRUE))
growth_factors <- growth_factors[growth_factors != "EFGLAM"]
df_beta <- mutate(df_beta, is_gf = hgnc_symbol %in% growth_factors)

ag_to_plot <- filter(df_beta, is_gf) %>%
  arrange(desc(c)) %>%
  head(n = 10) %>%
  .$hgnc_symbol

ag_to_plot <- c(ag_to_plot, "VEGFC")

ag_inds <- match(ag_to_plot, rowData(sce_gene)$hgnc_symbol)

marker_mat <- t(exprs(sce_gene))[, ag_inds, drop=FALSE]
colnames(marker_mat) <- ag_to_plot
marker_df <-  marker_mat %>% 
  as_data_frame() %>% 
  dplyr::mutate(pseudotime = tmap)

mx <- as.matrix(dplyr::select(marker_df, -pseudotime))
d <- dist(t(mx))
hc <- hclust(d)
gene_order <- rev(colnames(mx)[hc$order])

winsorize <- function(x, lower = -2.5, upper = 2.5) {
  x[x < lower] <- lower
  x[x > upper] <- upper
  x
}

marker_df_2 <- mutate(marker_df, pseudotime_order = rank(pseudotime)) %>% 
  gather(gene, expression, -pseudotime_order, -pseudotime)
marker_df_2$gene <- factor(marker_df_2$gene, levels = gene_order)

marker_df_2 <- group_by(marker_df_2, gene) %>% 
  mutate(norm_expression = winsorize((expression - mean(expression)) / sd(expression)))

ggplot(marker_df_2, aes(x = pseudotime_order, y = gene, fill = norm_expression)) +
         geom_raster(interpolate = FALSE) +
  scale_fill_viridis(name = "Expression\nz-score") +
  theme(legend.position = "right",
        axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 7, margin = margin(r = -0.5, l = 0, unit = "cm")),
        axis.text.x = element_blank(),
        axis.title.x = element_text(size = 10, margin = margin(t = 0, unit = "cm")),
        axis.ticks = element_blank(),
        axis.line = element_blank(),
        legend.title = element_text(size = 10),
        legend.text = element_text(size = 9),
        legend.direction = "vertical",
        legend.margin = margin(l = -.5, unit = "cm")) +
  labs(y = "Gene", x = "Pseudotime order")

heatmap_plot <- last_plot()
```





# Demanding supervisor plots

## Vascular growth factors

```{r vasc-growth-fact}
x_str <- plyr::mapvalues(x, from = sort(unique(x)),
                         to = c("ER-", "ER+"))
tmap <- sce$tmap

vgf <- c("VEGFA", "VEGFB", "VEGFC", "VEGFD", "FGF2", "CXCL8") 

mm <- match(vgf, rowData(sce_gene)$hgnc_symbol)

vgf_exprs <- t(exprs(sce_gene)[mm, ]) %>% as_data_frame()
names(vgf_exprs) <- vgf
vgf_exprs <- mutate(vgf_exprs, z = tmap, ER_status = x_str)
vgf_exprs_tidy <- gather(vgf_exprs, gene, expression, -z, -ER_status)

ggplot(vgf_exprs_tidy, aes(x = z, y = expression, color = ER_status)) +
  geom_point(alpha = 0.6) + facet_wrap(~ gene, scales = "free_y") +
  xlab("Pathway score") + ylab("Expression") +
  theme(legend.position = "top") +
  scale_color_brewer(palette = "Set1", name = "ER status") 

ggsave("../../figs/vgf_genes.png", width = 8, height = 6)
```

"Can we plot these brca genes?"

```{r brca-genes}
brca_genes <- c("CEP55", "ESR1", "FOXA1", "FOXC1", "KRT17", "MAPT", "MELK", "MMP11", "NAT1", "SFRP1", "UBE2C", "UBE2T")

mm <- match(brca_genes, rowData(sce_gene)$hgnc_symbol)

brca_exprs <- t(exprs(sce_gene)[mm, ]) %>% as_data_frame()
names(brca_exprs) <- brca_genes
brca_exprs <- mutate(brca_exprs, z = tmap, ER_status = x_str)
brca_exprs_tidy <- gather(brca_exprs, gene, expression, -z, -ER_status)

ggplot(brca_exprs_tidy, aes(x = z, y = expression, color = ER_status)) +
  geom_point(alpha = 0.6) + facet_wrap(~ gene, scales = "free_y") +
  xlab("Pathway score") + ylab("Expression") +
  theme(legend.position = "top") +
  scale_color_brewer(palette = "Set1", name = "ER status") 

ggsave("../../figs/brca_genes.png", width = 9, height = 6)
```

"Can we look at FBP1 vs SNAIL expression?"

```{r fbp1-vs-snail}
genes <- c("FBP1", "SNAI1")

mm <- match(genes, rowData(sce_gene)$hgnc_symbol)

fs_exprs <- t(exprs(sce_gene)[mm, ]) %>% as_data_frame()
names(fs_exprs) <- genes

fs_exprs <- mutate(fs_exprs, ER_status = x_str)

ggplot(fs_exprs, aes(x = FBP1, y = SNAI1, color = ER_status)) +
  geom_point(alpha = 0.6) +
  theme(legend.position = "top") +
  scale_color_brewer(palette = "Set1", name = "ER status") +
  xlab("FBP1 expression "~log[2]~"(TPM+1)") + 
  ylab(expression("SNAIL expression "~log[2]~"(TPM+1)"))

ggsave("../../figs/fbp1-vs-snail.png", width = 4, height = 4)
```

Boxplot for Chris:

```{r}
genes <- c("ESR1", "FBP1", "FOXC1")

mm <- match(genes, rowData(sce)$hgnc_symbol)

df_bp <- as_data_frame(t(exprs(sce)[mm, ]))
names(df_bp) <- genes
df_bp_tidy <- mutate(df_bp, ER_status = sce$ER_status) %>% 
  gather(gene, expression, -ER_status)

filter(df_bp_tidy, ER_status != "indeterminate") %>% 
  ggplot(aes(x = gene, y = expression, fill = ER_status)) +
    geom_violin() +
    scale_fill_brewer(palette = "Set1", name = "ER status") +
    labs(x = "Gene", y = "Normalised Expression")
ggsave("../../figs/brca_violinplot.png", width = 6.5, height = 4)
```


PCA experimenting:

```{r}
sce <- plotPCA(sce, ncomponents = 3, return_SCE = TRUE)

df_pca <- as_data_frame(reducedDim(sce)[,2:3]) %>% 
  mutate(pseudotime = sce$tmap, ER_status = sce$ER_status) %>% 
  filter(ER_status != "indeterminate")
df_pca$ER_status <- plyr::mapvalues(df_pca$ER_status,
                                    from = c("negative", "positive"),
                                    to = c("ER-negative", "ER-positive"))
df_pca$ER_status <- factor(df_pca$ER_status, levels = (unique(df_pca$ER_status)))
df_no_er <- select(df_pca, -ER_status)

df_curve <- frame_data(
  ~ ER_status, ~ x, ~ xend, ~ y, ~ yend, ~ curvature,
  "ER-positive", -50, 20, 27, 40, 0.1,
  "ER-negative", 55, 60, -35, 40, -0.1
)
df_curve$ER_status <- factor(df_curve$ER_status, levels = df_curve$ER_status)

plt <- filter(df_pca, ER_status != "indeterminate") %>% 
  ggplot(aes(x = PC2, y = PC3)) +
  geom_point(data = df_no_er, fill = "grey80", color = "grey80", size = 3) +
  geom_point(aes(fill = pseudotime), shape = 21, color = 'grey20', size = 3) +
  scale_fill_viridis(name = "Pseudotime", option = "C") + 
  facet_wrap(~ ER_status) +
  theme(legend.position = c(.4,.08),
        legend.direction = "horizontal") +
  geom_curve(aes(x = x, y = y, xend = xend, yend = yend),
             data = filter(df_curve, ER_status == "ER-negative"), color = 'black',
             curvature = df_curve$curvature[2], arrow = arrow(length = unit(0.3, "cm"), type = "open"), 
             size = 1.5) +
  geom_curve(aes(x = x, y = y, xend = xend, yend = yend),
             data = filter(df_curve, ER_status == "ER-positive"), color = 'black',
             curvature = df_curve$curvature[1], arrow = arrow(length = unit(0.3, "cm"), type = "open"), 
             size = 1.5) +
  theme(strip.background = element_rect(fill = "grey95", color = "grey30", size = 1),
        legend.text = element_text(size = 8),
        axis.text = element_text(size = 9),
        axis.title = element_text(size = 10),
        legend.title = element_text(size = 10)) +
  xlim(-72, 83) + ylim(-50, 50)
```


```{r}
library(grid)
genes <- c("FGF2", "FBP1", "FOXC1")
mm <- match(genes, rowData(sce)$hgnc_symbol)
gexp <- t(exprs(sce)[mm, ])
colnames(gexp) <- genes
gexp_df <- as_data_frame(gexp) %>% 
  mutate(pseudotime = sce$tmap)

df_bar <- inner_join(df_pca, gexp_df)

classify_cell <- function(row) {
  if(row['pseudotime'] > 0.5) return("end")
  if(row['pseudotime'] < -0.3 && row['ER_status'] == "ER-negative") return('beginning-negative')
  if(row['pseudotime'] < -0.3 && row['ER_status'] == "ER-positive") return("beginning-positive")
  return(NA)
}

df_bar$cell_class <- sapply(seq_len(nrow(df_bar)), function(i) classify_cell(df_bar[i, ]))

df_group <- filter(df_bar, !is.na(cell_class)) %>% 
  gather(gene, expression, -(PC2:ER_status), -cell_class) %>% 
  group_by(cell_class, gene) %>% 
  summarise(mean_expression = mean(expression))
df_group$gene <- factor(df_group$gene, levels = genes)

plts <- list()
for(cl in unique(df_group$cell_class)) {
  plts[[cl]] <- filter(df_group, cell_class == cl) %>% 
  ggplot(aes(x = gene, y = mean_expression, fill = gene)) +
  geom_bar(stat = "identity", color = 'grey20') +
  scale_fill_brewer(palette = "Dark2") +
  theme(legend.position = "None",
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 8, face = "bold"),
        axis.text.y = element_text(size = 8),
        axis.text.x = element_text(size = 8, angle = -90, face = "bold")) +
    labs(y = "Mean z-score\n expression") +
    ylim(0, 6.8)
}

width = 0.11
height = 0.3

pca_plot <- ggdraw() + 
  draw_plot(plt) +
  draw_plot(plts[["end"]], x = 0.4, y = 0.55, width = width, height = height) +
  draw_plot(plts[["end"]], x = 0.9, y = 0.55, width = width, height = height) +
  draw_plot(plts[["beginning-negative"]], x = 0.89, y = 0.14, width = width, height = height) +
  draw_plot(plts[["beginning-positive"]], x = 0.05, y = 0.555, width = width, height = height)



```


BRCA new figure 1:

```{r}
lsize = 12
bottom_grid <- plot_grid(heatmap_plot, goplot, 
                         rel_widths = c(3,2), label_size = lsize, nrow = 1,
                         labels = c("b", "c"))
grid <- plot_grid(pca_plot, bottom_grid, ncol = 1,
          labels = c("a", ""), label_size = lsize,
          rel_heights = c(3,2))
ggsave("../../figs/new/brca_1.png", grid, width = 10, height = 6, dpi = 600)

ggsave("../../figs/new/brca_1.svg", grid, width = 10, height = 6)
```

Save pseudotimes:

```{r}
df_save <- data_frame(
  sample_id = colnames(sce),
  tcga_barcode = colData(sce)$patient_barcode,
  pseudotime = pcavi$m_z
)

write_csv(df_save, "../../data/sample_pseudotimes_brca.csv")
```

